homeassistant/components/yamaha_musiccast/__init__.py

```diff
@@ -1 +1,171 @@
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    entry.add_update_listener(async_reload_entry)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Reload config entry."""
+    await async_unload_entry(hass, entry)
+    await async_setup_entry(hass, entry)
+
+
+class MusicCastDataUpdateCoordinator(DataUpdateCoordinator[MusicCastData]):
+    """Class to manage fetching data from the API."""
+
+    def __init__(self, hass: HomeAssistant, client: MusicCastDevice) -> None:
+        """Initialize."""
+        self.musiccast = client
+        self.platforms: list[str] = []
+        self.entities: list[Entity] = []
+
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL)
+
+    async def _async_update_data(self) -> MusicCastData:
+        """Update data via library."""
+        try:
+            await self.musiccast.fetch()
+            return self.musiccast.data
+        except Exception as exception:
+            raise UpdateFailed() from exception
+
+
+class MusicCastEntity(CoordinatorEntity):
+    """Defines a base MusicCast entity."""
+
+    def __init__(
+        self,
+        *,
+        entry_id: str,
+        coordinator: MusicCastDataUpdateCoordinator,
+        name: str,
+        icon: str,
+        enabled_default: bool = True,
+    ) -> None:
+        """Initialize the MusicCast entity."""
+        super().__init__(coordinator)
+        self._enabled_default = enabled_default
+        self._entry_id = entry_id
+        self._icon = icon
+        self._name = name
+        self._unsub_dispatcher = None
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def icon(self) -> str:
+        """Return the mdi icon of the entity."""
+        return self._icon
+
+    @property
+    def entity_registry_enabled_default(self) -> bool:
+        """Return if the entity should be enabled when first added to the entity registry."""
+        return self._enabled_default
+
+
+class MusicCastDeviceEntity(MusicCastEntity, abc.ABC):
+    """Defines a MusicCast device entity."""
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information about this MusicCast device."""
+        return DeviceInfo(
+            identifiers={
+                (
+                    DOMAIN,
+                    "".join(self.coordinator.data.mac_addresses.values()),
```

Comment:
We have a formatter for mac address:
https://github.com/home-assistant/core/blob/76edfe6c62c7ad49206813f2231ff79c4d99e2bb/homeassistant/helpers/device_registry.py#L116

Reply:
Didn't know that one, although the problem here is that a MusicCast device typically has more than one MAC address (wifi and eth) and we are joining those together to one big string. I'm unsure if that is the correct way to do

Reply:
We should preferably use the address that is used to connect to the device to Home Assistant. Do we know that?

Reply:
I checked the API and we can get the connection type and the mac address object is a dictionary by connection type. So yes we can get the mac for the currently used connection.
But, wouldn't that mean that when a user changes from WiFi to Ethernet for example that HA think it's a new device?

Reply:
Yes, we should probably use `connections` instead and set all the mac addresses. It's a set of tuples.

https://developers.home-assistant.io/docs/device_registry_index#device-properties

Reply:
Let's use the serial number in `identifiers`.